cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -2562,6 +2626,8 @@ Result<std::shared_ptr<io::OutputStream>> S3FileSystem::OpenOutputStream(
   ARROW_ASSIGN_OR_RAISE(auto path, S3Path::FromString(s));
   RETURN_NOT_OK(ValidateFilePath(path));
 
+  RETURN_NOT_OK(CheckS3Finalized());
```

Comment:
Can you put this consistently at the start of the method?

Reply:
Yes.
(I chose this position to use non aws-sdk-cpp related codes as much as possible even after `arrow::fs::FinalizeS3()`. But it may not be important in most cases.)